core/src/test/scala/integration/kafka/api/ProducerSendTestHelpers.scala

```diff
@@ -0,0 +1,238 @@
+          assertEquals(partition, metadata.partition)
+          offset += 1
+          timestampDiff += 1
+        } else {
+          fail(s"Send callback returns the following exception: $exception")
+        }
+      }
+    }
+
+    try {
+      // create topic
+      val topicProps = new Properties()
+      if (timestampType == TimestampType.LOG_APPEND_TIME)
+        topicProps.setProperty(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "LogAppendTime")
+      else
+        topicProps.setProperty(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "CreateTime")
+      TestUtils.createTopicWithAdmin(admin, topic, brokers, 1, 2, topicConfig = topicProps)
+
+      val recordAndFutures = for (i <- 1 to numRecords) yield {
+        val record = new ProducerRecord(topic, partition, baseTimestamp + i, s"key$i".getBytes(StandardCharsets.UTF_8),
+          s"value$i".getBytes(StandardCharsets.UTF_8))
+        (record, producer.send(record, callback))
+      }
+      producer.close(Duration.ofSeconds(20L))
+      recordAndFutures.foreach { case (record, future) =>
+        val recordMetadata = future.get
+        if (timestampType == TimestampType.LOG_APPEND_TIME)
+          assertTrue(recordMetadata.timestamp >= startTime && recordMetadata.timestamp <= System.currentTimeMillis())
+        else
+          assertEquals(record.timestamp, recordMetadata.timestamp)
+      }
+      assertEquals(numRecords, callback.offset, s"Should have offset $numRecords but only successfully sent ${callback.offset}")
+    } finally {
+      producer.close()
+    }
+  }
+
+  def defaultRecordAssertions(record: ConsumerRecord[Array[Byte], Array[Byte]], i: Int, now: Long, topic: String, partition: Int) = {
+    assertEquals(topic, record.topic)
+    assertEquals(partition, record.partition)
+    assertEquals(i.toLong, record.offset)
+    assertNull(record.key)
+    assertEquals(s"value${i + 1}", new String(record.value))
+    assertEquals(now, record.timestamp)
+  }
+
+
+  def sendToPartition(quorum: String, expectedRecords: Int = numRecords, recordAssertions: (ConsumerRecord[Array[Byte], Array[Byte]], Int, Long, String, Int) => Unit = defaultRecordAssertions _): Unit = {
+    val producer = createProducer()
+
+    try {
+      TestUtils.createTopicWithAdmin(admin, topic, brokers, 2, 2)
+      val partition = 1
+
+      val now = System.currentTimeMillis()
+      val futures = (1 to numRecords).map { i =>
+        producer.send(new ProducerRecord(topic, partition, now, null, ("value" + i).getBytes(StandardCharsets.UTF_8)))
+      }.map(_.get(30, TimeUnit.SECONDS))
+
+      // make sure all of them end up in the same partition with increasing offset values
+      for ((recordMetadata, offset) <- futures zip (0 until numRecords)) {
+        if (expectedRecords == numRecords) {
+          // This assertion only holds if we don't filter the records with an interceptor
+          assertEquals(offset.toLong, recordMetadata.offset)
+        }
```

Comment:
This is a bug caused by the fact the producer client always assumes that the number of `RecordMetadata` objects returned will match the number of records sent. This assumption doesn't hold when an interceptor filters out messages. Once that bug is fixed, the following should work:

```suggestion
      for ((recordMetadata, offset) <- futures zip (0 until expectedRecords)) {
        assertEquals(offset.toLong, recordMetadata.offset)
```